File: Inverted-Index/src/search/QueryParser.java
package search;

import java.util.ArrayList;
import java.util.List;

import keywords.Keyword;
import keywords.Ordinary;
import keywords.Exclude;
import keywords.Union;

public class QueryParser {
    private final String INCLUDE = "+";
    private final String EXCLUDE = "-";

    public List<Keyword> parseQuery(String query){

        var words = query.toLowerCase().split(" ");
        List<Keyword> keywords = new ArrayList<>();

        keywords.addAll(getOrdinaries(words));
        keywords.addAll(getUnions(words));
        keywords.addAll(getExcludes(words));
        
        return keywords;
    }
    public List<Keyword> getOrdinaries(String[] words) {

        var ordinaries = new ArrayList<Keyword>();
        for (var keyword : words) {
            if(!keyword.startsWith(INCLUDE) && !keyword.startsWith(EXCLUDE)) 
                ordinaries.add(new Ordinary(keyword));
        }
        return ordinaries;
    }
    public List<Keyword> getUnions(String[] words) {

        var additionals = new ArrayList<Keyword>();
        for (var keyword : words) {
            if(keyword.startsWith(INCLUDE)) 
                additionals.add(new Union(keyword.substring(1)));
        }
        return additionals;
    }
    public List<Keyword> getExcludes(String[] words) {

        var excludes = new ArrayList<Keyword>();
        for (var keyword : words) {
            if(keyword.startsWith(EXCLUDE)) 
                excludes.add(new Exclude(keyword.substring(1)));
        }
        return excludes;
    }
}

File: Phase5-TDD/InvertedIndex/Search/QueryParser.cs
using System.Collections.Generic;
using System;
using InvertedIndex.Keywords;

namespace InvertedIndex.Search
{
    public class QueryParser
    {
        private readonly string INCLUDE = "+";
        private readonly string EXCLUDE = "-";
        public IList<Keyword> ParseQuery(string query)
        {
            var words = query.ToLower().Split(" ");
            var keywords = new List<Keyword>();

            keywords.AddRange(getOrdinaries(words));
            keywords.AddRange(getUnions(words));
            keywords.AddRange(getExcludes(words));
            
            return keywords;
        }
        public IList<Keyword> getOrdinaries(string[] words)
        {
            var ordinaries = new List<Keyword>();
            foreach (var keyword in words) {
                if(!keyword.StartsWith(INCLUDE) && !keyword.StartsWith(EXCLUDE)) 
                    ordinaries.Add(new Ordinary(keyword));
            }
            return ordinaries;
        }
        public IList<Keyword> getUnions(string[] words)
        {
            var additionals = new List<Keyword>();
            foreach (var keyword in words) {
                if(keyword.StartsWith(INCLUDE)) 
                    additionals.Add(new Union(keyword.Substring(1)));
            }
            return additionals;
        }
        public IList<Keyword> getExcludes(string[] words)
        {
            var excludes = new List<Keyword>();
            foreach (var keyword in words) {
                if(keyword.StartsWith(EXCLUDE)) 
                    excludes.Add(new Exclude(keyword.Substring(1)));
            }
            return excludes;
        }
    }
}